airbyte-integrations/connectors/destination-kafka/src/main/java/io/airbyte/integrations/destination/kafka/KafkaRecordConsumer.java

```diff
@@ -0,0 +1,156 @@
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.KafkaException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class KafkaRecordConsumer extends FailureTrackingAirbyteMessageConsumer {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(KafkaRecordConsumer.class);
+
+  private final String topicPattern;
+  private final Map<AirbyteStreamNameNamespacePair, String> topicMap;
+  private final KafkaProducer<String, JsonNode> producer;
+  private final boolean sync;
+  private final boolean transactionalProducer;
+  private final ConfiguredAirbyteCatalog catalog;
+  private final Callback callback;
+  private final Consumer<AirbyteMessage> outputRecordCollector;
+  private final NamingConventionTransformer nameTransformer;
+
+  private AirbyteMessage lastStateMessage = null;
+
+  public KafkaRecordConsumer(KafkaDestinationConfig kafkaDestinationConfig,
+                             ConfiguredAirbyteCatalog catalog,
+                             Consumer<AirbyteMessage> outputRecordCollector,
+                             NamingConventionTransformer nameTransformer) {
+    this.topicPattern = kafkaDestinationConfig.getTopicPattern();
+    this.topicMap = new HashMap<>();
+    this.producer = kafkaDestinationConfig.getProducer();
+    this.sync = kafkaDestinationConfig.isSync();
+    this.transactionalProducer = kafkaDestinationConfig.isTransactionalProducer();
+    this.catalog = catalog;
+    this.callback = (metadata, exception) -> {
+      if (exception != null) {
+        // TODO improve error management
+        LOGGER.error("Error sending message to topic", exception);
+      }
+    };
+    this.outputRecordCollector = outputRecordCollector;
+    this.nameTransformer = nameTransformer;
+  }
+
+  @Override
+  protected void startTracked() {
+    Map<AirbyteStreamNameNamespacePair, String> mapped = catalog.getStreams().stream()
+        .map(stream -> AirbyteStreamNameNamespacePair.fromAirbyteSteam(stream.getStream()))
+        .collect(Collectors.toMap(Function.identity(),
+            pair -> nameTransformer.getIdentifier(topicPattern
+                .replaceAll("\\{namespace}", Optional.ofNullable(pair.getNamespace()).orElse(""))
+                .replaceAll("\\{stream}", Optional.ofNullable(pair.getName()).orElse("")))));
+    topicMap.putAll(mapped);
+
+    if (transactionalProducer) {
+      producer.initTransactions();
+    }
+  }
+
+  @Override
+  protected void acceptTracked(AirbyteMessage airbyteMessage) throws Exception {
+    if (airbyteMessage.getType() == AirbyteMessage.Type.STATE) {
+      lastStateMessage = airbyteMessage;
+    } else if (airbyteMessage.getType() == AirbyteMessage.Type.RECORD) {
+      final AirbyteRecordMessage recordMessage = airbyteMessage.getRecord();
+
+      final String topic = topicMap.get(AirbyteStreamNameNamespacePair.fromRecordMessage(recordMessage));
```

Comment:
what happens if these topics aren't already created in kafka? do we need to have particular permissions to create topics when using dynamic topic names?

Reply:
If the brokers have the property ``auto.create.topics.enable`` enabled (which is not recommended) these topics can be auto-created.
We could even use the AdminClient but I wouldn't do this